### ISSUE TITLE
Design and implementation

### ISSUE THREAD

**Issue by Rudxain** [Rick-Lang/rickroust#4]
# Intro
I'm not sure if this should be a [Discussion](https://github.com/Rick-Lang/rickroll-lang/discussions/79), but I believe it makes sense as an Issue.

I guess I should quote [The Purpose™](https://github.com/Rick-Lang/rickroll-lang/blob/7bae81bcfa5ed2f6e9b30ab82cea8d1a419f000d/README.md#purpose)
> an educational project for introducing various ways of implementing programming languages in a humorous and interactive way, without any painful processes of memorizing unfamiliar and complicated concepts!

which kinda serves as an informal "design-goals"/"philosophy" manifest of the Rick-Roll-Lang. Although it's not entirely clear how this affects the following sections, we should take it into account when making changes to the lang.

## Type system
Since this implementation will compile RR to machine code, I think it's worth considering [static-typing](https://en.wikipedia.org/wiki/Type_system#Static_type_checking) (with [inference](https://en.wikipedia.org/wiki/Type_inference), for ergonomics, of course)

RR has always had a [dynamic](https://en.wikipedia.org/wiki/Type_system#Dynamic_type_checking_and_runtime_type_information) sys (inherited from [Python 3](https://docs.python.org/3/reference/datamodel.html)), so perhaps we could go the TypeScript route and introduce [graduals](https://en.wikipedia.org/wiki/Gradual_typing).

I also find Gleam's type-sys kinda cool, especially the ~[parametric polymorphism](https://en.wikipedia.org/wiki/Parametric_polymorphism)~ [HM inference](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system), which (correct me if I misunderstood) allows function-param-types to be [inferred from the passed arguments on a per-call-site basis](https://github.com/gleam-lang/gleam/discussions/2713#discussioncomment-9034764) while rejecting any arg-type that's incompatible with the `fn` implementation in its body (thereby enforcing [soundness](https://en.wikipedia.org/wiki/Type_safety)), without the boilerplate of templates (see C++) or explicit type-parameters (see TS and Rust).

The Gleam Team is [considering type-classes](https://gleam.run/frequently-asked-questions/#will-gleam-have-type-classes), but IDK WTF they are :skull: (I know they are like Rust traits, but more general), so let's forget about that.

## Grammar
I believe we should take inspiration from Golang, while preserving most of the current syntax. I've been reading the [FAQ](https://go.dev/doc/faq#different_syntax) and the [cool semantics of `const`s](https://go.dev/blog/constants), and I've begun to have some appreciation for the lang. I'm still "married" to Rust, don't get me wrong lol, but Go is also a "respectable language".

As per the FAQ heading I linked, it should be obvious why Go is an appropriate template, as it's ["quite simple really"](https://youtu.be/7iHGoBVLtyY) (Mumbo Jumbo certified).

Even if we don't change the grammar/syntax at all, we could learn some stuff from `gc`'s parser, which is designed to be fast, but that's a secondary concern.

## Memory management
Clearly, the memory used by a RR program should be automatically-managed ([ref-count](https://en.wikipedia.org/wiki/Reference_counting) or [general GC](https://en.wikipedia.org/wiki/Garbage_collection_(computer_science))).

RR shouldn't have an ownership system (no "fighting the borrow-checker"), nor force the user to manually `alloc`ate and `free` memory. RR must be a fun lang!

But how do we implement it?:
- Does Inkwell take care of that?
- Should we `use` some crate/library that injects a GC runtime into built binaries?
- Should we reinvent the wheel?

# Outro
Anyways, thank y'all for your time. I hope this post is helpful

**Comment by SatinWukerORIG** [comment]
Basically, you have pointed out most of the problems of Rickroll in Rust. You can also examine this problem this way: we are rewriting a simple Python in Rust.

**Type System**: Parametric polymorphism is also the mechanism of Python's type system.
So, let's stick with it and implement a dynamic type for Rickroust.
This is actually doable in Inkwell, by dynamically checking variable types in runtime.

**Grammar**: we should preserve Rickroll's grammar because once we update it, we also have to change it for Rickroll in Python.
I am not worried about grammar; rather, I am concerned about whether we can implement this language efficiently and successfully.

**Memory Management**: I agree that we should have a more memory-free language because this is how Rickroll in Python is.
Let's use reference count because Inkwell supports low-level constructs like incrementing integers and conditional checks for freeing memory.
These are sufficient to implement reference counting without using additional tools.

**Goal**: Speaking of a design goal, my philosophy is: faster, simpler, and more maintainable. The faster and easier we can implement this the better. This does not mean that we don't care about the quality. In contrast, a fast and easy code base can help use maintain this repo and this language much better. Furthermore, Rickroll Language aims to make learners learn implementing languages in a "low-cost", efficient way. So, let's just keep everything simple!

**Comment by Rudxain** [comment]
> Parametric polymorphism is also the mechanism of Python's type system

That's true, but Py checks it at runtime while Gleam does it at compile-time (unless I'm confusing something? :confused:)

> implement a dynamic type for Rickroust.

I agree! Static-typed langs are usually designed for complex production projects, while RR is mostly a fun educational tool.

Replying to everything else: I agree, and I'm surprised that Inkwell has so many features!
